docs/proposals/004-initialization

```diff
@@ -0,0 +1,177 @@
+
+Related Work
+------------
+
+C++ has many different ways and syntax to initialize an object: through explicit constructor calls, initialization list, or implicitly in a member/variable declaration.
+A variable in C++ can also be in an uninitialized state after its declaration. HLSL inherits most of these behvior from C++ by allowing variables to be uninitialized.
+
+On the other hand, languages like C# and Swift has a set of well defined rules to ensure every variable is initialized after its declaration.
+
+Proposed Approach
+-----------------
+
+The high level direction for Slang is it needs to help ensuring all decls are properly initialized at its declaration site, while still providing a reasonable level
+of backward compatibility with existing Slang/HLSL code. In this section, we document all concepts and rules to achieve this goal.
+
+## `IDefaultInitializable` interface
+
+The builtin `IDefaultInitializable` interface is defined as:
+```csharp
+interface IDefaultInitializable
+{
+    __init();
+}
+```
+
+Any type that conforms to `IDefaultInitializable` is treated as having the *default-initializable* property in the rest of the discussions. By default, all builtin
+types, such as `int`, `float[]`, `float4` are default-initializable. The values for builtin numeric types after default-initialization are 0.
+The default value for pointers is `nullptr`, and default value for `Optional<T>` is `none`.
+
+## Default-Initializable Type
+
+A type X is default initializable if:
+- It explicitly declares that `X` implements `IDefaultInitializable`.
+- It explicitly provides a default constructor `X::__init()` that takes no arguments, in which case we treat the type as implementing `IDefaultInitializable` even if
+  this conformance isn't explicitly declared.
+- It is a struct type where all its members are default-initializable. A member is considered default-initializable if the type of the member is default-initializable,
+  or if the member has an initialization expression that defines its default value.
+- It is a sized-array type where the element type is default-initializable.
+- It is a tuple type where all element types are default-initializable.
+- It is an `Optional<T>` type for any `T`.
+
+This means that an unsized-array type or an existential type, or any composite type that contains such types without providing an explicit default constructor
+are not considered default-initializable.
+
+
+## Variable Initialization
+
+If the type of a local variable is default-initializable, then its default initializer will be invoked at its declaration site implicitly to intialize its value:
+```c++
+int x; // x will be default initialized to 0 because `int` is default-initializable.
+// The above is equivalent to:
+int x = int();
+
+struct S { int x; int y; }
+S s; // s will be default initialized to {0, 0} because `S` is default-initializable.
```

Comment:
Can we add few sentences to make it clear that

`S s;`
equal to 
`S s = S();`

And I have a question:
Is `S s = S()` equal to `S s = {}` ?

Reply:
Never mind, I see you you covered question this later.